### PR TITLE
set window title on loading a repo

### LINKED ISSUE #1691: Feature: use repo name as application (window) title?
**Is your feature request related to a problem? Please describe.**
Each of the applications in my taskbar has a title. For instance an instance of Firefox has the same title as the webpage that is currently open. So I have the icon to know that it's Firefox and I have the text that shows me that the loaded webpage is Github. This is useful, because sometimes you have multiple instance of an application opened and you need to distinguish them from one another. The lazygit application does not have a "title". The text by the icon is the path to the lazygit executable.

**Describe the solution you'd like**
I would like the lazygit application in my taskbar to display the name of the currently open repo.

**Describe alternatives you've considered**
None.

**Additional context**
I'm on Windows 10.
image

### PR DESCRIPTION
**Pull request by birgersp**
fix #1691

### CODE REVIEW

File: pkg/gui/gui.go [hunk 1]
```diff
@@ -778,3 +784,18 @@ func (gui *Gui) OnUIThread(f func() error) {
 		return f()
 	})
 }
+
+func (gui *Gui) UpdateWindowTitle() error {
```
Comment (jesseduffield) [inline pkg/gui/gui.go:788]: Could we put this function in `pkg/commands/oscommands/os.go`? We can then invoke it as `gui.OSCommand.UpdateWindowTitle()`.

File: pkg/gui/gui.go [hunk 2]
```diff
@@ -778,3 +784,18 @@ func (gui *Gui) OnUIThread(f func() error) {
...
+	if runtime.GOOS != "windows" {
+		return nil
+	}
+	path, getWdErr := os.Getwd()
+	if getWdErr != nil {
+		return getWdErr
+	}
+	cmd := exec.Command("cmd", "/C", "title", filepath.Base(path))
```
Comment (jesseduffield) [inline pkg/gui/gui.go:796]: I wonder if we should be saying 'RepoName - Lazygit' or 'Lazygit - RepoName' to make it clearer. Happy for you to make the call on this.

Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

**Review by jesseduffield — CHANGES_REQUESTED** [review]
Nice work! Left a couple comments

File: pkg/gui/gui.go (hunk 2 above)
Comment (jesseduffield) [reply pkg/gui/gui.go:796]: Another thing: we have an existing approach for running something in a shell that's platform-agnostic. That is, we could go:
```
func (c *OSCommand) SetWindowTitle(title string) error {
	return c.Cmd.NewShell("title " + c.Quote(title)).Run()
}
```
But I'm not on windows and don't know whether that works off-hand

File: pkg/gui/gui.go (hunk 1 above)
Comment (birgersp) [reply pkg/gui/gui.go]: Absolutely, I wasn't at all confident about that location

File: pkg/gui/gui.go (hunk 2 above)
Comment (birgersp) [reply pkg/gui/gui.go]: > I wonder if we should be saying 'RepoName - Lazygit' or 'Lazygit - RepoName' to make it clearer. Happy for you to make the call on this.
> 
> Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

From the other windows I have open it seems a bit mixed, but Google Chrome and VSCode both have the "[custom title] - [application name]" format so I think maybe that makes most sense.

Comment (birgersp) [reply pkg/gui/gui.go]: > Another thing: we have an existing approach for running something in a shell that's platform-agnostic. That is, we could go:
> 
> ```
> func (c *OSCommand) SetWindowTitle(title string) error {
> 	return c.Cmd.NewShell("title " + c.Quote(title)).Run()
> }
> ```
> 
> But I'm not on windows and don't know whether that works off-hand

Makes sense, but how would I invoke it from [gui.loadNewRepo](https://github.com/jesseduffield/lazygit/blob/master/pkg/gui/gui.go#L677)?

Comment (birgersp) [reply pkg/gui/gui.go]: Nevermind, I got it :)

> Also we'll need to ensure that if our repo has spaces in the name that this command still works as expected.

Good thinking. Turns out that we need to write out the title without quotation marks, even when there are spaces.

So the command is like this: `title This is my window title`.

image

**Comment by birgersp** [comment]
@jesseduffield Idk if I did this PR process correctly (I probably shouldn't have amended and force pushed when doing changes, I'll keep that in mind going forward). Anyways, I have addressed your comments.

File: pkg/commands/oscommands/os.go [hunk 3]
```diff
@@ -267,3 +268,16 @@ func GetLazygitPath() string {
...
+	if runtime.GOOS != "windows" {
+		return nil
+	}
+	path, getWdErr := os.Getwd()
+	if getWdErr != nil {
+		return getWdErr
+	}
+	title := fmt.Sprint(filepath.Base(path), " - Lazygit")
+	args := append([]string{"cmd", "/C", "title"}, strings.Split(title, " ")...)
+	return c.Cmd.NewShell(strings.Join(args, " ")).Run()
```
Comment (jesseduffield) [inline pkg/commands/oscommands/os.go:282]: if I check the output in the Command Log panel I see `bash -c "cmd /C title lazygit - Lazygit"`. I'm on osx so it uses bash but I suspect on windows it shows `cmd /C "cmd /C title lazygit - Lazygit`. In effect, the NewShell function adds the `cmd` and `/C` arguments itself. So I'd see how you go removing those two args here

Comment (birgersp) [reply pkg/commands/oscommands/os.go]: Thanks! I've removed the redundant args. Also I noticed that the OSCommand struct has os/platform information so I removed the dependency to the "runtime" package.

**Comment by birgersp** [comment]
@jesseduffield you wrote

> if I check the output in the Command Log panel I see `bash -c "cmd /C title lazygit - Lazygit"`. I'm on osx so it uses bash (snip)

I wouldn't expect you to see any output in the Command Log panel from this, because of this condition:

```
	if c.Platform.OS != "windows" {
		return nil
	}
```

Is the `c.Platform.OS` property really "windows" in your debugging session (on OSX), too?

**Comment by jesseduffield** [comment]
@birgersp sorry I didn't mention that I removed that check for the sake of seeing what would be logged

**Comment by jesseduffield** [comment]
Code looks good. If you want to add a test feel free, but you'd need to store `getwd` on the OSCommand struct as a dependency so we can mock it out in the test. I'm not going to force you to go through that process if you don't want to! 

Running CI now

**Comment by jesseduffield** [comment]
Merged. Thanks for your effort @birgersp !

**Comment by birgersp** [comment]
Great! Thank you!

> Code looks good. If you want to add a test feel free, but you'd need to store `getwd` on the OSCommand struct as a dependency so we can mock it out in the test. I'm not going to force you to go through that process if you don't want to!
> 
> Running CI now

Would like to add a test, but I'm a complete beginner in Go so it's all a bit new for me. But I can try.

**Pull request merged**